### PR TITLE
feat: resize oversize images before sending

### PR DESCRIPTION
**Pull request by SkyTNT**
添加自动缩小过大图片功能

### CODE REVIEW

File: packages/core/src/index.ts [hunk 1]
```diff
@@ -97,22 +98,35 @@ class ImageService extends Service {
...
+      let contentType = resp.headers.get('content-type')
+      if (resize) {
+        let img = sharp(buffer)
+        const metadata = await img.metadata()
+        const ratio = size / Math.max(metadata.width, metadata.height)
+        if(ratio < 1){
+          img = img.resize({
+            width: Math.floor(metadata.width * ratio),
+            height: Math.floor(metadata.height * ratio),
+          })
```
Comment (Lipraty) [inline packages/core/src/index.ts:114]: Do not delegate other error to the network layer; they should be handled separately. You are trying to catch errors here that do not originate from the network request layer.

File: packages/core/src/index.ts
Comment (Lipraty) [inline packages/core/src/index.ts:1]: Network requests have a detailed process, and changes here can be disruptive. Are you tested this correctly?

File: packages/core/src/index.ts [hunk 2]
```diff
@@ -97,22 +98,35 @@ class ImageService extends Service {
...
-        return `data:${resp.headers['content-type']};base64,${Buffer.from(resp.data).toString('base64')}`
-      })
-      .catch((err) => {
-        if (Quester.Error.is(err)) {
-          logger.warn(
-            `Request images failed with HTTP status ${err.response?.status}: ${JSON.stringify(err.response?.data)}.`,
-          )
-        } else {
-          logger.error(`Request images failed with unknown error: ${err.message}.`)
+  async imgUrlToBase64(url: string, resize: boolean = false, size: number = 1280): Promise<string> {
```
Comment (Lipraty) [inline packages/core/src/index.ts:101]: You are performing the resizing operation in an incorrect function.

File: packages/core/src/command.tsx [hunk 3]
```diff
@@ -83,14 +83,15 @@ export function apply(ctx: Context, config: Config) {
         }
         url ||= image.url
 
-        if (config.asset && ctx.assets) {
-          url = await ctx.booru.imgUrlToAssetUrl(url)
+        if (config.base64 || config.autoResize) {
```
Comment (Lipraty) [inline packages/core/src/command.tsx:86]: What is your plan if the target platform does not support sending base64 images? (e.g., adapter-qq)

File: package.json [hunk 4]
```diff
@@ -48,6 +48,7 @@
     "mocha": "^9.2.2",
     "prettier": "^3.2.5",
     "sass": "^1.62.1",
+    "sharp": "^0.33.4",
```
Comment (Lipraty) [inline package.json:51]: This package is a Node addon; please manage its side effects properly.

**Review by Lipraty — CHANGES_REQUESTED** [review]

File: packages/core/src/index.ts [hunk 5]
```diff
@@ -97,22 +98,35 @@ class ImageService extends Service {
...
+        let img = sharp(buffer)
+        const metadata = await img.metadata()
+        const ratio = size / Math.max(metadata.width, metadata.height)
+        if (ratio < 1) {
+          img = img.resize({
+            width: Math.floor(metadata.width * ratio),
+            height: Math.floor(metadata.height * ratio),
+          })
+          buffer = await img.jpeg({ quality: 80 }).toBuffer()
+          contentType = 'image/jpeg'
```
Comment (SaarChaffee) [inline packages/core/src/index.ts:116]: 不应该直接写在 `imgUrlToBase64` 中，这破坏了函数原本的职责

File: packages/core/src/command.tsx [hunk 6]
```diff
@@ -83,14 +83,15 @@ export function apply(ctx: Context, config: Config) {
...
+          url = await ctx.booru.imgUrlToBase64(url, config.autoResize, preferSizesToSize[config.preferSize])
```
Comment (SaarChaffee) [inline packages/core/src/command.tsx:87]: 不要强行走 `imgUrlToBase64`，在大部分情况下这会造成非必要的性能浪费

File: packages/core/src/index.ts [hunk 7]
```diff
@@ -97,22 +98,35 @@ class ImageService extends Service {
...
+      return `data:${contentType};base64,${buffer.toString('base64')}`
+    } catch (err) {
+      if (Quester.Error.is(err)) {
+        logger.warn(
+          `Request images failed with HTTP status ${err.response?.status}: ${JSON.stringify(err.response?.data)}.`,
+        )
+      } else {
+        logger.error(`Request images failed with unknown error: ${err.message}.`)
+      }
+      return null
```
Comment (SaarChaffee) [inline packages/core/src/index.ts:128]: 这里只处理了 `http` 的异常

File: packages/core/src/index.ts [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import { Context, Logger, Quester, Schema, Service, remove } from 'koishi'
 import LanguageDetect from 'languagedetect'
+import sharp from 'sharp'
```
[the remark this answers is not in the document]
Comment (SkyTNT) [reply packages/core/src/index.ts:3]: How to manage it.

**Review by SaarChaffee — CHANGES_REQUESTED** [review]

**Comment by Lipraty** [comment]
Thank you very much for your contribution. Here are some issues I noticed:

- There is no management of side effects for the package using the Node addon, which can disrupt Koishi's HMR and subsequent update operations.
- The resizing method should be independent rather than coupled with the toBase64 method.
- Please respect the original coding style.

Some thoughts on potential solutions:

- For packages with strong platform dependencies like Node addon/gyp, try to avoid using them and instead opt for packages written purely in JavaScript or manage the Node addon.
- toBase64 is just one method of sending images for booru and should not be forcibly coupled.

Reference:

https://koishi.chat/zh-CN/cookbook/design/storage.html#%E5%BA%94%E5%BD%93%E9%81%BF%E5%85%8D%E7%9A%84%E6%83%85%E5%86%B5

https://forum.koishi.xyz/t/topic/1930

File: packages/core/src/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Lipraty) [reply packages/core/src/index.ts:3]: I investigated the sharp package, which is a gyp package that compiles C++ code into a Node addon during installation. This problematic approach makes it difficult to manage its side effects and disrupts Koishi's update and HMR processes. Therefore.

I recommend switching to another package that can manage side effects (_directly require on `.node` files_). This way, you can fork and change it and use the `downloads` plugin for  download `.node` file separately.

[the remark this answers is not in the document]
Comment (MaikoTan) [reply packages/core/src/index.ts:3]: Or just use some pure JS implementation like [jimp](https://www.npmjs.com/package/jimp).

**Comment by SaarChaffee** [comment]
另外，有部分 api 支持选择不同尺寸或大小的图片。如果有，应当优先使用 api 所提供的。

**Comment by SkyTNT** [comment]
> 另外，有部分 api 支持返回不同尺寸或大小的图片。如果有，应当优先使用 api 所提供的。

原本逻辑应该就是优先使用 api 所提供的大小吧。我只是加了如果超过选择的大小就自动缩放，这是可选的。

File: packages/core/src/source.ts [hunk 9]
```diff
@@ -91,3 +91,4 @@ export namespace ImageSource {
 }
 
 export const preferSizes = ['thumbnail', 'large', 'medium', 'small', 'original'] as const
+export const preferSizesToSize = { thumbnail: 128, large: 1280, medium: 640, small: 320, original: 2048 } as const
```
Comment (Lipraty) [inline packages/core/src/source.ts:94]: Why does the `original` need to be size-limited?

**Review by Lipraty — CHANGES_REQUESTED** [review]

File: packages/core/src/source.ts (hunk 9 above)
Comment (SkyTNT) [reply packages/core/src/source.ts:94]: Well, the original should not be limited.

**Comment by SkyTNT** [comment]
现在再看看怎么样

File: packages/core/src/index.ts [hunk 10]
```diff
@@ -101,7 +130,7 @@ class ImageService extends Service {
     return this.ctx
       .http(url, { method: 'GET', responseType: 'arraybuffer', proxyAgent: '' })
       .then((resp) => {
-        return `data:${resp.headers['content-type']};base64,${Buffer.from(resp.data).toString('base64')}`
+        return `data:${resp.headers.get('content-type')};base64,${Buffer.from(resp.data).toString('base64')}`
```
Comment (MaikoTan) [inline packages/core/src/index.ts:133]: This change can be split in a separated fixup PR.

Other changes are still in consideration.

**Comment by SkyTNT** [comment]
jimp存在内存和性能问题，考虑再看看有没有其它可以用的库
https://github.com/jimp-dev/jimp/issues/153
https://github.com/jimp-dev/jimp/issues/915

**Comment by SkyTNT** [comment]
明天试试用canvas能不能实现

**Comment by MaikoTan** [comment]
> 明天试试用canvas能不能实现

Use canvas might be a better idea since many users have it installed.

**Comment by MaikoTan** [comment]
By the way, you should also add this option into the document, in this case, it would go to the page https://github.com/koishijs/koishi-plugin-booru/blob/main/docs/zh-CN/config.md

**Comment by SkyTNT** [comment]
现在应该可以了吧

**Review by SaarChaffee — APPROVED** [review]

**Comment by SkyTNT** [comment]
等等，似乎还要优化一下，我看puppeteer的 toBuffer是基于 toDataURL实现的，多了不必要的base64转换

**Comment by SkyTNT** [comment]
应该可以了

**Review by MaikoTan — APPROVED** [review]
Now it should be good to merge

@Lipraty make a double check please.

**Pull request merged**